src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,531 @@
+
+                var nonce = GenerateRandomBytes();
+
+                var document = new BsonDocument()
+                    .Add("r", new BsonBinaryData(nonce))
+                    .Add("p", new BsonInt32('n'));
+
+                var clientFirstMessageBytes = ToBytes(document);
+
+                return new ClientFirst(clientFirstMessageBytes, nonce, _credential, _sessionToken, _dateTimeProvider);
+            }
+
+            private byte[] GenerateRandomBytes()
+            {
+                return _randomByteGenerator.Generate(ClientNonceLength);
+            }
+        }
+
+        private class ClientFirst : ISaslStep
+        {
+            private readonly byte[] _bytesToSendToServer;
+            private readonly UsernamePasswordCredential _credential;
+            private readonly IClock _dateTimeProvider;
+            private readonly byte[] _nonce;
+            private readonly string _sessionToken;
+
+            public ClientFirst(
+                byte[] bytesToSendToServer,
+                byte[] nonce,
+                UsernamePasswordCredential credential,
+                string sessionToken,
+                IClock dateTimeProvider)
+            {
+                _bytesToSendToServer = bytesToSendToServer;
+                _credential = credential;
+                _dateTimeProvider = dateTimeProvider;
+                _nonce = nonce;
+                _sessionToken = sessionToken;
+            }
+
+            public byte[] BytesToSendToServer
+            {
+                get { return _bytesToSendToServer; }
+            }
+
+            public bool IsComplete
+            {
+                get { return false; }
+            }
+
+            public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedFromServer)
+            {
+                var serverFirstMessageDoc = ToDocument(bytesReceivedFromServer);
+                var serverNonce = serverFirstMessageDoc["s"].AsByteArray;
+                var host = serverFirstMessageDoc["h"].AsString;
+
+                if (serverNonce.Length != ClientNonceLength * 2 || !serverNonce.Take(ClientNonceLength).SequenceEqual(_nonce))
+                {
+                    throw new MongoAuthenticationException(conversation.ConnectionId, message: "Server sent an invalid nonce.");
+                }
+
+                AwsSignatureVersion4.SignRequest(
+                    _dateTimeProvider.UtcNow,
+                    _credential.Username,
+                    _credential.GetInsecurePassword(),
+                    _sessionToken,
+                    serverNonce,
+                    host,
+                    out var authHeader,
+                    out var timestamp);
+
+                var document = new BsonDocument()
+                    .Add("a", authHeader)
+                    .Add("d", timestamp);
+
+                if (_sessionToken != null)
+                {
+                    document.Add("t", _sessionToken);
+                }
```

Comment:
Lines 491-498 would be written more idiomatically as:

    var document = new BsonDocument
    {
        { "a", authHeader },
        { "d", timestamp },
        { "t",  _sessionToken, _sessionToken != null }
    };